### PR TITLE
verilog: keep trace with instance in TAGS file

### PR DESCRIPTION
**Pull request by my2817**
add a kind named K_INST
update function tagNameList to update kind from K_IDENTIFIER to K_INST

### CODE REVIEW

**Comment by masatake** [comment]
The way to write test cases is wrong. expected.tags is missing. See https://docs.ctags.io/en/latest/units.html.
You should not use `_` as the part of the kind name. The tests on the CI environment may be failed because of the `_` in the kind name.

**Comment by hirooih** [comment]
Thanks for enhancement.

- Please use terminology used in[ the LRM (Language Reference Manual)](https://ieeexplore.ieee.org/document/8299595), inst -> instance.
- Verilog does not support interface.
- Please check with the following tests.

```
make units LANGUAGES=SystemVerilog,Verilog
make units LANGUAGES=SystemVerilog,Verilog VG=1
```

- Make sure only instance names are tagged.

**Comment by hirooih** [comment]
> Make sure only instance names are tagged.

And;

- Make sure all instance names are tagged.

**Comment by my2817** [comment]
> > Make sure only instance names are tagged.
> 
> And;
> 
> * Make sure all instance names are tagged.

Thanks for your work on this awesome tool and advice.

File: parsers/verilog.c
Comment (masatake) [inline parsers/verilog.c:1585]: we don't use this style, putting `{` at the end of line IN THIS FILE.

```
if (c == '(')
{
   ...
```
is expected.

File: parsers/verilog.c
Comment (masatake) [inline parsers/verilog.c:1589]: we don't use this style, putting { at the end of line IN THIS FILE.

if (c == '(')
{
   ...

is expected.

File: parsers/verilog.c
Comment (masatake) [inline parsers/verilog.c:1593]: we don't use this style, putting { at the end of line IN THIS FILE.

else
{
   ...

is expected.

File: parsers/verilog.c
Comment (my2817) [reply parsers/verilog.c:1589]: Thanks, it's my first time to do this job. I will back again

File: parsers/verilog.c
Comment (masatake) [inline parsers/verilog.c:1594]: How about printing the kind in a string?
```
verbose("find instance: %s with kind %s\n", vStringValue (token->name),getNameForKind(token->kind));
```

File: parsers/verilog.c
Comment (masatake) [reply parsers/verilog.c:1589]: This project is looking for persons like you!

File: parsers/verilog.c
Comment (masatake) [reply parsers/verilog.c:1594]: (I myself didn't test the change that I showed.)

**Comment by masatake** [comment]
I would like you to squash the commits into one commit.
The idea behind the squashing is described in https://github.com/universal-ctags/ctags/pull/2705#issuecomment-730041445 .

In addition, I would like you to use "SystemVerilog" or "Verilog,SystemVerilog" instead of "[verilog/SystemVerilog]" as the prefix of the commit header like: 

```
cc26cc1d Ruby: fill signature fields for the method kind
65f5bc5b Go: use symbol table for attaching typeref field to members of a structure
bf1206cc SystemVerilog: parse a multi dimensional packed array in typedef correctly
```

File: parsers/verilog.c
Comment (masatake) [inline parsers/verilog.c:1590]: We can simplify the code a bit like:
```
if (c == '(')
{
      kind = K_UNDEFINED;
      c = skipPastMatch ("()");
      if (c == ';' || c == ',')
      {
			kind = K_INSTANCE;
            verbose("find instance: %s with kind %s\n", vStringValue (token->name), getNameForKind(kind));        
      }
      token->kind = kind;
}
```

Comment (hirooih) [reply parsers/verilog.c:1590]: I will check this code in detail in this week.
Quick comments.

I propose the followings; 

```
if (c == '(')
{
      c = skipPastMatch ("()");
      if (c == ';' || c == ',')
      {
		kind = K_INSTANCE;
        	verbose("find instance: %s with kind %s\n", vStringValue (token->name), getNameForKind(kind));        
      }
      else  // comment why this condition is required
		kind = K_UNDEFINED;

      token->kind = kind;
}
```

Actually I don't understand why the condition is required. 

From LRM 23.3.2 Module instantiation syntax

```
module_instantiation ::= // from A.4.1.1
	module_identifier [ parameter_value_assignment ] hierarchical_instance { , hierarchical_instance };
...
hierarchical_instance ::= name_of_instance ( [ list_of_port_connections ] )
name_of_instance ::= instance_identifier { unpacked_dimension }
...
```

Please take care of ` { unpacked_dimension }` part.

Please check interface instantiation syntax, too.

Comment (my2817) [reply parsers/verilog.c]: From the units test, the parser will treat "`add_t" as instance if there isn't a "else" condition
``` verilog
    var `add_t(foo) = '0;
```
I will check unpacked_dimension and interface syntax later, thanks.

File: parsers/verilog.c
Comment (hirooih) [inline parsers/verilog.c:1591]: We need `c= skipWhite (c);` here.

`m inst ()  ; // a space before semicolon` fails.   (You may want to add a test case for this. )

`skipPastMatch()` should do the job, but it does not now.  I will work on this issue on this weekend.
You may wait for the fix.

File: parsers/verilog.c
Comment (hirooih) [inline parsers/verilog.c:132]: Please be consistent with the line above.  i.e. `"instance (module, instance)"`

File: parsers/verilog.c
Comment (hirooih) [inline parsers/verilog.c:145]: ditto

**Review by hirooih — CHANGES_REQUESTED** [review]
Thank you for your proposal.

Instance name support is very important. I completely missed the issue.
Please see my comments.

There is one issue.  If I understand your fix correctly, it tags a void function as a instance.

```
// LRM 13.4.1 Return values and void functions
myprint(a); // call myprint (defined below) as a statement
function void myprint (int a);  // FIXME
...
endfunction
```

Void function is not used often.  Supporting instance tag is much more important.  Leave it as known bug for now.

Please add the test case above, too.

Thanks.

File: parsers/verilog.c
Comment (hirooih) [inline parsers/verilog.c:1583]: Remove a space you added.

File: parsers/verilog.c
Comment (hirooih) [inline parsers/verilog.c:1604]: This else clause is not necessary.

File: parsers/verilog.c
Comment (hirooih) [inline parsers/verilog.c:1593]: This if-clause is not necessary.

File: parsers/verilog.c
Comment (hirooih) [inline parsers/verilog.c:1599]: This variable is not necessary.

File: parsers/verilog.c
Comment (hirooih) [inline parsers/verilog.c:1593]: add a space after 'if'.

File: parsers/verilog.c
Comment (hirooih) [inline parsers/verilog.c:132]: If you chose "instances of module or interface" below, this should be "instances of module".

File: parsers/verilog.c
Comment (hirooih) [inline parsers/verilog.c:145]: Add a space before "}".

**Review by hirooih — CHANGES_REQUESTED** [review]
Sorry for many change requests...

File: parsers/verilog.c
Comment (my2817) [reply parsers/verilog.c:1599]: OK, I'm sorry for my poor coding skill 😭

File: parsers/verilog.c
Comment (hirooih) [inline parsers/verilog.c:1643]: It seems that you missed my review comment this morning.
This variable is not necessary.  You can specify K_INSTANCE directly.

File: parsers/verilog.c
Comment (hirooih) [inline parsers/verilog.c:1640]: I've sent a pull-request #2718 for this.  I am waiting for the approval by @masatake.
If it is merged earlier than this pull-request, remove this line.

Again add a test case which fails by this issue.  (Add a space in ");" at the end of one of module instantiations.)

**Review by hirooih — CHANGES_REQUESTED** [review]
It is getting simpler, isn't it?

File: parsers/verilog.c
Comment (my2817) [reply parsers/verilog.c:1643]: Got it, thanks!

File: parsers/verilog.c
Comment (hirooih) [inline parsers/verilog.c:1646]: `kind` is in this line, too.

**Review by hirooih — CHANGES_REQUESTED** [review]
See my comment.

> Again add a test case which fails by this issue. (Add a space in ");" at the end of one of module instantiations.)

For example;
verilog-instance.d/input.v: L57

```
             .b                         (b));
```

Change this line as;

```
             .b                         (b))   ;  // spaces before semicolon
```

Remove;
```
			c = skipWhite (c);
```

and confirm that the test fails.

And #2718 is merged into `universal-ctags:master`.
Merge it into your branch and confirm the test passes.

**Comment by hirooih** [comment]
One more thing.
Have you checked the coverage report?  (It is a really awesome tool.)

https://codecov.io/gh/universal-ctags/ctags/compare/4a9f00a1c199133729819043292aedff7c580ed8...2069733b92a881ec74bd5419d186611e7432e870/changes

Your fix made the following lines unnecessary.

```
1643 | 1660 |   | /* skip port list of module instance: foo bar(xx, yy); */
1644 | 1661 |   | c = skipWhite (c);
1645 | 1662 |   | if (c == '(')
1646 | 1663 |   | {
1647 | 1664 |   | c = skipPastMatch ("()");
1648 | 1665 |   | c = skipWhite (c);
1649 | 1666 |   | }
```

Please remove them.

Your fix added very small number of lines and gave us great functionality.  It sounds great!

**Review by hirooih — APPROVED** [review]

**Comment by hirooih** [comment]
Again, thank you for your contribution.

**Comment by my2817** [comment]
> 
> 
> Again, thank you for your contribution.

Thanks for your patience!

**Comment by masatake** [comment]
I'm working on VHDL and I have studied some concepts of this technology area.
I got some ideas to improve SystemVerilog parser.
```
module foo # (parameter
...
endmodule: foo
...

  foo uut7 [1:0][10:0]();
```

for uut7, ctags with this pull request emits:
```
uut7   input.v /^   foo uut7 [1:0][10:0]();$/;"        i       module:top
```

I think we can fill typeref: field for representing the relationship between uut7 and foo.
```
uut7   input.v /^   foo uut7 [1:0][10:0]();$/;"        i       module:top       typeref:module:foo
```

I'm taking the same approach in VHDL.

**Comment by hirooih** [comment]
Thanks.
I've added the feature on the TODO list #2674 .

**Pull request merged**
